examples/ecs/custom_query_param.rs

```diff
@@ -0,0 +1,165 @@
+use bevy::{
+    ecs::{
+        component::Component,
+        query::{Fetch, FilterFetch},
+    },
+    prelude::*,
+};
+use std::{fmt::Debug, marker::PhantomData};
+
+/// This examples illustrates the usage of `Fetch` and `FilterFetch` derive macros, that allow
+/// defining custom query and filter types.
+///
+/// White regular tuple queries work great in most of simple scenarios, using custom queries
+/// declared as named structs can bring the following advantages:
+/// - They help to avoid destructuring or using `q.0, q.1, ...` access pattern
+/// - Adding, removing components or changing items order with structs greatly reduces maintenance
+///   burden, as you don't need to update statements that destructure tuples, care abort order
+///   of elements, etc. Instead, you can just add or remove places where a certain element is used
+/// - Named structs enable the composition pattern, that makes query types easier to re-use
+/// - They allow to go over the limit of 15 components that exists for query tuples
+///
+/// For more details on the `Fetch` and `FilterFetch` derive macros, see their documentation.
+fn main() {
+    App::new()
+        .add_startup_system(spawn)
+        .add_system(print_components.system().label("print_components"))
+        .add_system(
+            print_components_readonly
+                .system()
+                .label("print_components_readonly")
+                .after("print_components"),
+        )
+        .add_system(
+            print_components_tuple
+                .system()
+                .after("print_components_readonly"),
+        )
+        .run();
+}
+
+#[derive(Debug)]
+struct ComponentA;
+#[derive(Debug)]
+struct ComponentB;
+#[derive(Debug)]
+struct ComponentC;
+#[derive(Debug)]
+struct ComponentD;
+#[derive(Debug)]
+struct ComponentZ;
+
+#[derive(Fetch)]
+struct MutQuery<'w, T: Component + Debug, P: Component + Debug> {
+    entity: Entity,
+    a: Mut<'w, ComponentA>,
```

Comment:
Please specify here, that `Mut<'w, ComponentA>` is the necessary replacement for `&mut ComponentA`.